File: frontend/lib/models/user_model.dart
class SubscriptionDetails {
  final String? subscriptionId;
  final String? planId;
  final String? status;
  final DateTime? currentPeriodEnd;

  SubscriptionDetails({
    this.subscriptionId,
    this.planId,
    this.status,
    this.currentPeriodEnd,
  });

  factory SubscriptionDetails.fromJson(Map<String, dynamic>? json) {
    if (json == null) {
      return SubscriptionDetails(); // Return default/empty if no subscription data
    }
    return SubscriptionDetails(
      subscriptionId: json['subscriptionId'] as String?,
      planId: json['planId'] as String?,
      status: json['status'] as String?,
      currentPeriodEnd:
          json['currentPeriodEnd'] != null
              ? DateTime.tryParse(json['currentPeriodEnd'] as String)
              : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'subscriptionId': subscriptionId,
      'planId': planId,
      'status': status,
      'currentPeriodEnd': currentPeriodEnd?.toIso8601String(),
    };
  }
}

class UserModel {
  final String id;
  final String name;
  final String email;
  final List<String> favorites;
  final List<String> watchHistory; // Added watch history
  final String role;
  final String? googleId;
  final String? stripeCustomerId; // New field
  final SubscriptionDetails? subscription; // New field

  UserModel({
    required this.id,
    required this.name,
    required this.email,
    required this.favorites,
    required this.watchHistory, // Added to constructor
    required this.role,
    this.googleId,
    this.stripeCustomerId, // Added to constructor
    this.subscription, // Added to constructor
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['_id'] as String? ?? json['id'] as String? ?? '',
      name:
          json['name'] as String? ??
          json['username'] as String? ??
          '', // Try both name and username fields
      email: json['email'] as String? ?? '',
      favorites: List<String>.from(json['favorites'] as List? ?? []),
      watchHistory: List<String>.from(
        json['watchHistory'] as List? ?? [],
      ), // Added parsing
      role: json['role'] as String? ?? 'customer',
      googleId: json['googleId'] as String?,
      stripeCustomerId:
          json['stripeCustomerId'] as String?, // Parsing new field
      subscription:
          json['subscription'] != null
              ? SubscriptionDetails.fromJson(
                json['subscription'] as Map<String, dynamic>,
              )
              : null, // Parsing new field
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {
      '_id': id,
      'username': name,
      'email': email,
      'favorites': favorites,
      'watchHistory': watchHistory, // Added to JSON
      'role': role,
    };
    if (googleId != null) {
      data['googleId'] = googleId;
    }
    if (stripeCustomerId != null) {
      data['stripeCustomerId'] = stripeCustomerId; // Adding new field
    }
    if (subscription != null) {
      data['subscription'] = subscription!.toJson(); // Adding new field
    }
    return data;
  }
}

File: backend/models/user.model.js
import mongoose from "mongoose";
import bcrypt from "bcryptjs";

const userSchema = new mongoose.Schema(
  {
    username: {
      type: String,
      required: [true, "Name is required"],
      unique: true,
      trim: true,
      minlength: 3,
    },

    email: {
      type: String,
      required: [true, "Email is required"],
      unique: true,
      trim: true,
      lowercase: true,
    },

    googleId: {
      type: String,
      unique: true,
      sparse: true,
    },

    password: {
      type: String,
      required: [
        function () {
          return !this.googleId;
        },
        "Password is required",
      ],
      minlength: [6, "Password must be at least 6 characters"],
      maxLength: 128,
      select: false,
      trim: true,
    },

    favorites: [
      {
        type: String,
      },
    ],

    watchHistory: [
      {
        videoId: {
          type: String,
          required: true,
        },
        watchedAt: {
          type: Date,
          default: Date.now,
        },
      },
    ],

    stripeCustomerId: {
      type: String,
    },
    subscription: {
      subscriptionId: {
        type: String,
      },
      planId: {
        type: String,
      },
      status: {
        type: String,
        enum: [
          "active",
          "canceled",
          "past_due",
          "inactive",
          "incomplete",
          "incomplete_expired",
          "trialing",
          "unpaid",
          "paused",
        ],
        default: "inactive",
      },
      currentPeriodEnd: {
        type: Date,
        set: function (date) {
          if (!date) {
            return new Date(Date.now() + 30 * 24 * 60 * 60 * 1000); // Default to 30 days from now
          }
          if (typeof date === "number") {
            return new Date(date * 1000);
          }
          if (date instanceof Date) {
            return date;
          }
          // Try to parse string date
          const parsedDate = new Date(date);
          if (!isNaN(parsedDate.getTime())) {
            return parsedDate;
          }
          // If all else fails, default to 30 days from now
          return new Date(Date.now() + 30 * 24 * 60 * 60 * 1000);
        },
      },
    },

    role: {
      type: String,
      enum: ["customer", "admin"],
      default: "customer",
    },
  },

  {
    timestamps: true,
  }
);

userSchema.pre("save", async function (next) {
  if (!this.isModified("password")) {
    return next();
  }
  try {
    const salt = await bcrypt.genSalt(10);
    this.password = await bcrypt.hash(this.password, salt);
    next();
  } catch (error) {
    next(error);
  }
});

userSchema.methods.matchPassword = async function (password) {
  return await bcrypt.compare(password, this.password);
};

const User = mongoose.model("User", userSchema);

export default User;
